analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/InitializationVectorShouldBeRandomBase.cs

```diff
@@ -18,10 +18,108 @@
+        {
+            return ProcessStrongRandomGeneratorMethodInvocation(invocation, state)
+                    ?? ProcessCreateEncryptorMethodInvocation(invocation, state, context)
+                    ?? state;
+        }
+        else
+        {
+            return state;
+        }
+    }
+
+    private static ProgramState ProcessAssignmentToIVProperty(IAssignmentOperationWrapper assignment, ProgramState state)
+    {
+        return assignment.Target?.AsPropertyReference() is { } property
+                && GetIVPropertyInstance(property) is { } instanceSymbol
+                && (assignment.Value.AsArrayCreation() is { } || state[assignment.Value].HasConstraint(ByteCollectionConstraint.CryptographicallyWeak))
+                    ? state.SetSymbolConstraint(instanceSymbol, ByteCollectionConstraint.CryptographicallyWeak)
+                    : null;
+
+        static ISymbol GetIVPropertyInstance(IPropertyReferenceOperationWrapper property) =>
+            property.Property.Name.Equals("IV")
+            && property.Instance.TrackedSymbol() is { } instanceSymbol
+            && instanceSymbol.GetSymbolType().DerivesFrom(KnownType.System_Security_Cryptography_SymmetricAlgorithm)
+               ? instanceSymbol
+               : null;
+    }
+
+    private static ProgramState ProcessStrongRandomGeneratorMethodInvocation(IInvocationOperationWrapper invocation, ProgramState state)
+    {
+        if (IsCryptographicallyStrongRandomNumberGenerator(invocation)
+            && invocation.ArgumentValue("data") is { } byteArray
+            && byteArray.TrackedSymbol() is { } byteArraySymbol)
+        {
+            return state.SetSymbolConstraint(byteArraySymbol, ByteCollectionConstraint.CryptographicallyStrong);
+        }
+        else if (IsGenerateIVMethod(invocation))
+        {
+            return state.SetSymbolConstraint(invocation.Instance.TrackedSymbol(), ByteCollectionConstraint.CryptographicallyStrong);
+        }
+        else
+        {
+            return null;
+        }
+    }
+
+    private ProgramState ProcessCreateEncryptorMethodInvocation(IInvocationOperationWrapper invocation, ProgramState state, SymbolicContext context)
+    {
+        if (IsCreateEncryptorMethod(invocation)
+            && (SecondArgumentIsCryptographicallyWeakArray(invocation, state) || HasCryptographicallyWeakIVProperty(invocation, state)))
+        {
+            ReportIssue(context.Operation.Instance, invocation.Instance.Syntax.ToString());
+        }
+        return state;
+
+        bool SecondArgumentIsCryptographicallyWeakArray(IInvocationOperationWrapper invocation, ProgramState state) =>
+            invocation.Arguments.Length == 2
+            && IvArgument(invocation) is { } ivArgument
+            && (ivArgument.AsArrayCreation() is { } || (ivArgument.TrackedSymbol() is { } symbol && state[symbol]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is true));
```

Comment:
I think we should be way less specific here. I don't think we get any unexpected cases here. But even if we did, passing something `CryptographicallyWeak` would be a reason for concern. And the code becomes much simpler this way.

```suggestion
        bool ArgumentIsCryptographicallyWeak(IInvocationOperationWrapper invocation, ProgramState state) =>
            invocation.Arguments.Any(x => state[x]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is true);
```